Deprecate `qiskit.Aer` and `qiskit.providers.aer`

jakelishman opened the pull request:
### Summary

These were intended to be deprecated after the release of Aer 0.11, but we apparently never actually got around to it.  This fully deprecates the `qiskit.Aer` object, along with imports from `qiskit.providers.aer`, as we will be removing those completely in Qiskit 1.0.

The namespace loaders in `qiskit.namespace` will also be removed in Qiskit 1.0, but I have not marked these with deprecation warnings because:
- we need to continue using them to effect the deprecations of the imports.
- they have never been documented as public interface, so are a private implementation detail.

All the use of `qiskit.providers.aer` in Terra and its tests is replaced with `qiskit_aer`.  This name is only available from Aer 0.11, but in practice, Aer 0.10 already does not work with Qiskit 0.46 (fails during import), so this change does not affect the working set of Aer versions.

I haven't touched old imports of IBMQ, because that package is already completely deprecated and we should be removing all attempted use of it from Qiskit 1.0, because that code would never be able to succeed.

<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Details and comments

See #5089 for the history of these deprecations.

jakelishman committed:
Deprecate `qiskit.Aer` and `qiskit.providers.aer`

These were intended to be deprecated after the release of Aer 0.11, but
we apparently never actually got around to it.  This fully deprecates
the `qiskit.Aer` object, along with imports from `qiskit.providers.aer`,
as we will be removing those completely in Qiskit 1.0.

The namespace loaders in `qiskit.namespace` will also be removed in
Qiskit 1.0, but I have not marked these with deprecation warnings
because:
- we need to continue using them to effect the deprecations of the
  imports.
- they have never been documented as public interface, so are a private
  implementation detail.

All the use of `qiskit.providers.aer` in Terra and its tests is replaced
with `qiskit_aer`.  This name is only available from Aer 0.11, but in
practice, Aer 0.10 already does not work with Qiskit 0.46 (fails during
import), so this change does not affect the working set of Aer versions.

I haven't touched old imports of IBMQ, because that package is already
completely deprecated and we should be removing all attempted use of it
from Qiskit 1.0, because that code would never be able to succeed.

jakelishman committed:
Fix lint

ElePT commented on qiskit/__init__.py:
```diff
@@ -145,40 +145,39 @@ def __init__(self):
+            warnings.warn(
+                "The qiskit.IBMQ entrypoint and the qiskit-ibmq-provider package ("
+                "accessible from 'qiskit.providers.ibmq`) are deprecated and will be removed "
+                "in a future release. Instead you should use the qiskit-ibm-provider package "
```
The `qiskit-ibm-provider` repo is being deprecated and the provider is being migrated to `qiskit-ibm-runtime`. It might be good to keep track of this PR and modify the message accordingly https://github.com/Qiskit/qiskit-ibm-runtime/pull/1285. Because I am not super sure of the timing, we could also keep this warning but I don't think it's ideal to point to another deprecated package.

jakelishman replied:
Ah thanks, I didn't know the timings on that at all.  Let's track that separately, because this warning isn't actually new for this PR - I just moved it so that the warning triggers when you _attempt_ the import, regardless of whether it succeeds or not.

jakelishman replied:
Tracked in #11450.

ElePT commented on qiskit/providers/fake_provider/fake_backend.py:
```diff
@@ -554,11 +554,11 @@ def run(self, run_input, **kwargs):
-                from qiskit.providers import aer
-                from qiskit.providers.aer.pulse import PulseSystemModel
+                import qiskit_aer
+                from qiskit_aer.pulse import PulseSystemModel
```
Note https://github.com/Qiskit/qiskit/pull/11442#discussion_r1433917371

jakelishman replied:
Added a deprecation warning in b299aae.

jakelishman committed:
Deprecate running pulse jobs on fake backends

With Aer 0.13 removing the capability, this is not something we will be
able to support moving forwards.  This deprecates the functionality, in
preparation for its full removal in Qiskit 1.0.

mtreinish commented on qiskit/test/decorators.py:
```diff
@@ -55,14 +54,7 @@ def is_aer_provider_available():
     Returns:
         bool: True if simulator executable is available
     """
-    # TODO: HACK FROM THE DEPTHS OF DESPAIR AS AER DOES NOT WORK ON MAC
```
How long as this been here? Aer has worked on mac for a really long time now...

jakelishman replied:
```
commit b2bbb3f05968f3cadc2993ed66123077d80026ef
Author: Jay M. Gambetta <jay.gambetta@us.ibm.com>
Date:   Fri Dec 28 17:32:27 2018 -0500

    Remove legacy simulators  (#1615)
```

so not too long, you know

mtreinish commented on releasenotes/notes/deprecate-aer-hooks-1bc00a96ebb269f6.yaml:
```diff
@@ -0,0 +1,14 @@
+    should instead use the same object from the ``qiskit_aer`` namespace, which is a drop-in
+    replacement.
+  - |
+    Importing from ``qiskit.providers.aer`` is deprecated and will stop working in Qiskit 1.0.  You
```
Did we have a deprecation warning for the namespace hooks in general? Not that it hurts to do this explicitly because we're deprecating the backwards compat import machinery in this PR.

jakelishman replied:
What kind of statements do you mean, sorry?

If you mean the objects in `qiskit.namespace` themselves, I haven't put deprecations on them a) we need to use them for the whole length of their existence so would trigger warnings by our regular use of them (or we'd have to wait another cycle to deprecate), and b) they're not public per our policy (there's no documentation of `qiskit.namespace`).

jakelishman replied:
For linking purposes: at the time, we discussed whether to make `qiskit.namespace` explicitly private, but equally didn't make a positive decision that it would be _public_ per the deprecation policy, we just didn't call it `qiskit._namespace`.  By [the intended current terms of the deprecation policy](https://github.com/Qiskit/documentation/pull/366/), it doesn't count as public because it's not documented, though caveat: we never merged #11205 on Terra, and then the deprecation policy got moved to `DEPRECATIONS.md` on Terra in #11218 in its old form, so we need to resurrect #11205 here.

mtreinish replied:
I meant https://github.com/Qiskit/qiskit/pull/11223 which implicitly covered `qiskit.providers.aer` at least in my head, the `qiskit.namespace` machinery was just an explicit exception to that.

mtreinish commented on tools/find_optional_imports.py:
```diff
@@ -29,7 +29,7 @@ def _main():
         "ipywidgets",
         "scipy.stats",
         "matplotlib",
-        "qiskit.providers.aer",
```
We should probably have both TBH as either path will work. While it's probably pretty unlikely that someone will import `qiskit.providers.aer` in a backport PR protecting against it doesn't add any real overhead.

jakelishman replied:
Sounds sensible, done in ef656370c.

mtreinish reviewed (commented):
Overall this LGTM. At the time we added the namespace import redirect I think we discussed just leaving `qiskit.providers.aer` -> `qiskit_aer` in perpetuity because it didn't have any real cost just the `qiskit.namespace` module. I think removing it for Qiskit 1.0.0 makes sense though if we're going to do it.

Just a couple small comments inline.

jakelishman committed:
Restore old namespace checking to optionals finder

jakelishman commented:
I'm keen to get rid of `qiskit.namespace` because I'm not convinced it's going to be maintainable properly through new Python versions - I think some of the methods we implement are already removed in Python 3.12 (`pylint` throws new exceptions due to missing docstrings that were previously inherited on 3.12), and I know there's more changes planned for `importlib` in 3.13/3.14.

mtreinish commented:
That does raise a concern for me in rustworkx where we use the same basic code to redirect imports from `retworkx` for backwards compatibility there. We're not ready to remove that redirect since we're planning to drop it in an eventual rustworkx 1.0.0 release (which is only being loosely planned for at this point).

jakelishman committed:
Correct deprecation warning message

jakelishman commented:
To be fair, I'm 100% certain it'll still be possible to do this, I just think we'll have to go look at the `importlib` documentation and work out the new preferred way of doing it.  I was keen to avoid doing that unless we needed to on Qiskit, but if you need it on Rustworkx, I guess it's a little unavoidable.

Fixed the dodgy deprecation message in 4add053d7a.
